bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java

```diff
@@ -0,0 +1,508 @@
+                    .locatorFor(delegate.getPolyglotContext(), engineIdentifier, scriptExtensionAccessor)
+                    .locateModule(name, fromlist);
+            delegate.getBindings(ScriptContext.ENGINE_SCOPE).put(ScriptExtensionModuleProvider.IMPORT_PROXY_NAME,
+                    wrapImportFn);
+            try {
+                String wrapperContent = new String(
+                        Files.readAllBytes(PythonScriptEngineFactory.PYTHON_WRAPPER_FILE_PATH));
+                delegate.getPolyglotContext().eval(Source.newBuilder(GraalPythonScriptEngine.LANGUAGE_ID,
+                        wrapperContent, PythonScriptEngineFactory.PYTHON_WRAPPER_FILE_PATH.toString()).build());
+
+                // inject scope, Registry and logger
+                if (!pythonScriptEngineConfiguration.isInjection(PythonScriptEngineConfiguration.INJECTION_DISABLED)
+                        && (ctx.getAttribute("javax.script.filename") == null || pythonScriptEngineConfiguration
+                                .isInjection(PythonScriptEngineConfiguration.INJECTION_ENABLED_FOR_ALL_SCRIPTS))) {
+                    String injectionContent = "import scope\nfrom openhab import Registry, logger";
+                    delegate.getPolyglotContext().eval(Source
+                            .newBuilder(GraalPythonScriptEngine.LANGUAGE_ID, injectionContent, "<generated>").build());
+                }
+            } catch (IOException e) {
+                throw new IllegalArgumentException("Failed to generate import wrapper", e);
+            }
+        }
+
+        // logger initialization, for non file based scripts, has to be delayed, because ruleUID is not available yet
+        if (ctx.getAttribute("javax.script.filename") == null) {
+            Runnable wrapperLoggerFn = () -> setScriptLogger();
+            delegate.getBindings(ScriptContext.ENGINE_SCOPE).put(LOGGER_INIT_NAME, wrapperLoggerFn);
+        } else {
+            setScriptLogger();
+        }
+
+        initialized = true;
+    }
+
+    @Override
+    protected String beforeInvocation(String source) {
+        source = super.beforeInvocation(source);
+
+        // Happens for Transform and UI based rules (eval and compile)
+        // and has to be evaluate every time, because of changing and late injected ruleUID
+        if (delegate.getBindings(ScriptContext.ENGINE_SCOPE).get(LOGGER_INIT_NAME) != null) {
+            return LOGGER_INIT_NAME + "()\n" + source;
+        }
+
+        return source;
+    }
+
+    @Override
+    protected Object afterInvocation(Object obj) {
+        lock.unlock();
+        logger.debug("Lock released after invocation.");
+        return super.afterInvocation(obj);
+    }
+
+    @Override
+    protected Exception afterThrowsInvocation(Exception e) {
+        // OPS4J Pax Logging holds a reference to the exception, which causes the OpenhabGraalJSScriptEngine to not be
```

Comment:
Probably update my comment:

```suggestion
        // OPS4J Pax Logging holds a reference to the exception, which causes the PythonScriptEngine to not be
```

Reply:
fixed